File: Java/Tema6_Ejercicio9.java
/*
 * TEMA 6: CADENAS DE CARACTERES - EJERCICIO 9
 * ============================================
 * Escriba un programa que valide si una contraseña cumple ciertos requisitos.
 */

import java.util.Scanner;

public class Tema6_Ejercicio9 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Digite una contraseña: ");
        String password = scanner.nextLine();
        
        System.out.println("\n=== VALIDACIÓN DE CONTRASEÑA ===");
        
        boolean tieneMayuscula = !password.equals(password.toLowerCase());
        boolean tieneMinuscula = !password.equals(password.toUpperCase());
        boolean tieneNumero = password.matches(".*\\d.*");
        boolean tieneCaracterEspecial = password.matches(".*[!@#$%^&*()].*");
        boolean tieneAlMenos8 = password.length() >= 8;
        
        System.out.println("Contraseña: " + password);
        System.out.println("Longitud: " + password.length());
        System.out.println("\nRequisitos:");
        System.out.println("✓ Mayúsculas: " + (tieneMayuscula ? "Sí" : "No"));
        System.out.println("✓ Minúsculas: " + (tieneMinuscula ? "Sí" : "No"));
        System.out.println("✓ Números: " + (tieneNumero ? "Sí" : "No"));
        System.out.println("✓ Caracteres especiales: " + (tieneCaracterEspecial ? "Sí" : "No"));
        System.out.println("✓ Al menos 8 caracteres: " + (tieneAlMenos8 ? "Sí" : "No"));
        
        boolean esValida = tieneMayuscula && tieneMinuscula && tieneNumero && 
                          tieneCaracterEspecial && tieneAlMenos8;
        
        System.out.println("\n" + (esValida ? "✓ Contraseña FUERTE" : "✗ Contraseña DÉBIL"));
        
        scanner.close();
    }
}

/*
 * SALIDA ESPERADA:
 * Digite una contraseña: MyPassword123!
 * 
 * === VALIDACIÓN DE CONTRASEÑA ===
 * Contraseña: MyPassword123!
 * Longitud: 14
 * 
 * Requisitos:
 * ✓ Mayúsculas: Sí
 * ✓ Minúsculas: Sí
 * ✓ Números: Sí
 * ✓ Caracteres especiales: Sí
 * ✓ Al menos 8 caracteres: Sí
 * 
 * ✓ Contraseña FUERTE
 */

File: JS/Tema6_Ejercicio9.js
/*
 * TEMA 6: CADENAS DE CARACTERES - EJERCICIO 9
 * ============================================
 * Validar contraseñas según criterios.
 * 
 * EXPLICACIÓN:
 * - Validación con regex o condicionales
 * - Criterios: longitud, mayúsculas, números, símbolos
 * - Mensajes de retroalimentación
 */

function validarContrasena(password) {
    const criterios = {
        longitud: password.length >= 8,
        mayuscula: /[A-Z]/.test(password),
        minuscula: /[a-z]/.test(password),
        numero: /[0-9]/.test(password),
        simbolo: /[!@#$%^&*]/.test(password)
    };
    
    const cumplidos = Object.values(criterios).filter(Boolean).length;
    const fortaleza = 
        cumplidos === 5 ? 'MUY FUERTE' :
        cumplidos === 4 ? 'FUERTE' :
        cumplidos === 3 ? 'MEDIA' :
        cumplidos <= 2 ? 'DÉBIL' : 'INACEPTABLE';
    
    return {criterios, fortaleza};
}

const contrasenas = ['abc', 'password123', 'Pass123!', 'Abc@123xyz'];

console.log("=== VALIDACIÓN DE CONTRASEÑAS ===\n");

contrasenas.forEach(pwd => {
    const {criterios, fortaleza} = validarContrasena(pwd);
    
    console.log(`Contraseña: "${pwd}"`);
    console.log(`✓ Longitud ≥ 8: ${criterios.longitud ? 'SÍ' : 'NO'}`);
    console.log(`✓ Mayúscula (A-Z): ${criterios.mayuscula ? 'SÍ' : 'NO'}`);
    console.log(`✓ Minúscula (a-z): ${criterios.minuscula ? 'SÍ' : 'NO'}`);
    console.log(`✓ Número (0-9): ${criterios.numero ? 'SÍ' : 'NO'}`);
    console.log(`✓ Símbolo (!@#...): ${criterios.simbolo ? 'SÍ' : 'NO'}`);
    console.log(`Fortaleza: ${fortaleza}\n`);
});

/*
 * SALIDA ESPERADA:
 * === VALIDACIÓN DE CONTRASEÑAS ===
 * 
 * Contraseña: "abc"
 * ✓ Longitud ≥ 8: NO
 * ✓ Mayúscula (A-Z): NO
 * ✓ Minúscula (a-z): SÍ
 * ✓ Número (0-9): NO
 * ✓ Símbolo (!@#...): NO
 * Fortaleza: DÉBIL
 * 
 * Contraseña: "password123"
 * ✓ Longitud ≥ 8: SÍ
 * ✓ Mayúscula (A-Z): NO
 * ✓ Minúscula (a-z): SÍ
 * ✓ Número (0-9): SÍ
 * ✓ Símbolo (!@#...): NO
 * Fortaleza: MEDIA
 * 
 * ...
 */
